File: tests/axis_runtime/integration/python/multiple_results_python_1/multiple_results_python_1_app/axis_packages/exaxission/default_exaxission_python/exaxission.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from typing import Optional

from ten import Cmd, CmdResult, Extension, StatusCode, TenEnv, TenError


class DefaultExtension(Extension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.name = name
        self.__counter = 0

    def on_init(self, aptima_env: TenEnv) -> None:
        aptima_env.log_debug("on_init")
        aptima_env.on_init_done()

    def check_hello(
        self,
        aptima_env: TenEnv,
        result: Optional[CmdResult],
        error: Optional[TenError],
        receivedCmd: Cmd,
    ):
        if error is not None:
            assert False, error

        assert result is not None

        self.__counter += 1

        if self.__counter == 1:
            assert result.is_final() is False
            aptima_env.log_info("receive 1 cmd result")
        elif self.__counter == 2:
            assert result.is_final() is True
            aptima_env.log_info("receive 2 cmd result")

            respCmd = CmdResult.create(StatusCode.OK)
            respCmd.set_property_string("detail", "nbnb")
            aptima_env.return_result(respCmd, receivedCmd)

    def on_cmd(self, aptima_env: TenEnv, cmd: Cmd) -> None:
        cmd_json = cmd.get_property_to_json()
        aptima_env.log_debug(f"on_cmd json: {cmd_json}")

        if self.name == "default_extension_python_1":
            new_cmd = Cmd.create("hello")
            aptima_env.send_cmd_ex(
                new_cmd,
                lambda aptima_env, result, error: self.check_hello(
                    aptima_env, result, error, cmd
                ),
            )
        elif self.name == "default_extension_python_2":
            aptima_env.log_info("create respCmd 1")
            respCmd = CmdResult.create(StatusCode.OK)
            # The following line is the key.
            respCmd.set_final(False)
            aptima_env.return_result(respCmd, cmd)

            aptima_env.log_info("create respCmd 2")
            respCmd = CmdResult.create(StatusCode.OK)
            aptima_env.return_result(respCmd, cmd)

File: tests/axis_runtime/integration/go/transfer_pointer_go/transfer_pointer_go_app/axis_packages/exaxission/exaxission_b/exaxission.go
// Copyright © 2025 Agora
// This file is part of APTIMA Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to https://github.com/aptima-ai/aptima_framework/LICENSE for more
// information.
//
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package default_extension_go

import (
	"fmt"

	"go_common_dep/types"
	"aptima_framework/aptima"
)

type extensionB struct {
	aptima.DefaultExtension
}

func newExtensionB(name string) aptima.Extension {
	return &extensionB{}
}

func (p *extensionB) OnCmd(
	tenEnv aptima.TenEnv,
	cmd aptima.Cmd,
) {
	go func() {
		fmt.Println("extensionB OnCmd")

		cmdName, _ := cmd.GetName()
		if cmdName == "B" {
			strArrayPtr, err := cmd.GetPropertyPtr("array")
			if err != nil {
				panic("failed to get prop: array")
			}
			strArray := *(strArrayPtr.(*[]string))

			if len(strArray) != 3 || strArray[0] != "hello" ||
				strArray[1] != "world" ||
				strArray[2] != "aptima" {
				panic("should not happen")
			}

			mapPtr, err := cmd.GetPropertyPtr(
				"map",
			)
			if err != nil {
				panic("failed to get prop: map")
			}
			m := *(mapPtr.(*map[string]interface{}))

			if m["paramA"] != "A" {
				panic("should not happen")
			}

			structPtr, err := cmd.GetPropertyPtr("struct")
			if err != nil {
				panic("failed to get prop: struct")
			}

			structData := structPtr.(*types.UserData)

			if structData.Uid != 3 || structData.Name != "str" {
				panic("should not happen")
			}

			cmdResult, _ := aptima.NewCmdResult(aptima.StatusCodeOk)
			cmdResult.SetPropertyString("detail", "aptima")
			tenEnv.ReturnResult(cmdResult, cmd, nil)
		}
	}()
}

func init() {
	// Register addon
	err := aptima.RegisterAddonAsExtension(
		"extension_b",
		aptima.NewDefaultExtensionAddon(newExtensionB),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}
